File: src/erlang/algorithms_1.erl
-module(two_sum).
-export([two_sum/2]).

-spec two_sum(Nums :: [integer()], Target :: integer()) -> [integer()].
two_sum(Nums, Target) ->
    IndexNum = lists:zip(lists:seq(0, length(Nums)-1), Nums),

    Pairs = [{I, J} || 
            {I, X} <- IndexNum , 
            {J, Y} <- IndexNum ,  
            I < J , 
            X + Y =:= Target],

    case Pairs of
        [{I, J} | _] -> [I, J];
        [] -> {}
    end
  .

File: src/cpp/algorithms_1.cpp
#include <unordered_map>
#include <vector>
using namespace std;

class Solution {
  public:
    vector<int> twoSum(vector<int>& nums, int target) {
        using Value = int;
        using Index = int;
        std::unordered_map<Value, Index> mp;

        for (int i = 0; i < nums.size(); i++) {
            if (mp.count(target - nums[i])) {
                return {i, mp[target - nums[i]]};
            }
            mp[nums[i]] = i;
        }

        return {};
    }
};

